### PR TITLE
Fix batch progress state misalignment when resuming from mid-epoch checkpoints

### LINKED ISSUE #19367: Potential off by 1 error when resuming training of mid-epoch checkpoint
### Bug description

During the fit loop, here's a simple log of the `global_step` and `batch_idx` values during `on_train_batch_start` and `on_train_batch_end`.
```
[TRAIN STEP START] trainer.global_step=0, batch_idx=0
[TRAIN STEP END] trainer.global_step=1, batch_idx=0
[TRAIN STEP START] trainer.global_step=1, batch_idx=1
[TRAIN STEP END] trainer.global_step=2, batch_idx=1
[TRAIN STEP START] trainer.global_step=2, batch_idx=2
[TRAIN STEP END] trainer.global_step=3, batch_idx=2
[TRAIN STEP START] trainer.global_step=3, batch_idx=3
[TRAIN STEP END] trainer.global_step=4, batch_idx=3
[TRAIN STEP START] trainer.global_step=4, batch_idx=4
[TRAIN STEP END] trainer.global_step=5, batch_idx=4
[VAL STEP START] trainer.global_step=5, batch_idx=0
[VAL STEP START] trainer.global_step=5, batch_idx=1
[VAL STEP START] trainer.global_step=5, batch_idx=2
[VAL STEP START] trainer.global_step=5, batch_idx=3
[VAL STEP START] trainer.global_step=5, batch_idx=4
[TRAIN STEP START] trainer.global_step=5, batch_idx=5
[TRAIN STEP END] trainer.global_step=6, batch_idx=5
[TRAIN STEP START] trainer.global_step=6, batch_idx=6
[TRAIN STEP END] trainer.global_step=7, batch_idx=6
[TRAIN STEP START] trainer.global_step=7, batch_idx=7
[TRAIN STEP END] trainer.global_step=8, batch_idx=7
[TRAIN STEP START] trainer.global_step=8, batch_idx=8
[TRAIN STEP END] trainer.global_step=9, batch_idx=8
[TRAIN STEP START] trainer.global_step=9, batch_idx=9
[TRAIN STEP END] trainer.global_step=10, batch_idx=9
[VAL STEP START] trainer.global_step=10, batch_idx=0
[VAL STEP START] trainer.global_step=10, batch_idx=1
[VAL STEP START] trainer.global_step=10, batch_idx=2
[VAL STEP START] trainer.global_step=10, batch_idx=3
[VAL STEP START] trainer.global_step=10, batch_idx=4
`Trainer.fit` stopped: `max_steps=10` reached.
```
Notice that global_step and batch_idx are equal during batch_start and global step is 1 greater than batch index for batch_end. Now, if I save a mid-epoch checkpoint after 5 training steps and resume training, I see the following
```
[TRAIN STEP START] trainer.global_step=5, batch_idx=4
[TRAIN STEP END] trainer.global_step=6, batch_idx=4
[VAL STEP START] trainer.global_step=6, batch_idx=0
[VAL STEP START] trainer.global_step=6, batch_idx=1
[VAL STEP START] trainer.global_step=6, batch_idx=2
[VAL STEP START] trainer.global_step=6, batch_idx=3
[VAL STEP START] trainer.global_step=6, batch_idx=4
[TRAIN STEP START] trainer.global_step=6, batch_idx=5
[TRAIN STEP END] trainer.global_step=7, batch_idx=5
[TRAIN STEP START] trainer.global_step=7, batch_idx=6
[TRAIN STEP END] trainer.global_step=8, batch_idx=6
[TRAIN STEP START] trainer.global_step=8, batch_idx=7
[TRAIN STEP END] trainer.global_step=9, batch_idx=7
[TRAIN STEP START] trainer.global_step=9, batch_idx=8
[TRAIN STEP END] trainer.global_step=10, batch_idx=8
`Trainer.fit` stopped: `max_steps=10` reached.
```
Now the two values are off by 1 during batch start and off by 2 during batch end. This seems to be an issue because it changes when validation and checkpointing is run. In both runs, I have `Trainer(val_check_interval=5, ...)` and `ModelCheckpoint(every_n_train_steps=5, ...)`. In the original run, validation happens after 5 and 10 training steps, as expected. In the resumed run, validation only happens once after 6 training steps.

My initial guess is that this is happening because `self.batch_progress.increment_completed()` in `src/lightning/pytorch/loops/training_epoch_loop.py` is called after
```
call._call_callback_hooks(trainer, "on_train_batch_end", batch_output, batch, batch_idx)
call._call_lightning_module_hook(trainer, "on_train_batch_end", batch_output, batch, batch_idx)
trainer._logger_connector.on_batch_end()
```
so the checkpoint thinks we've only completed `global_steps-1` training steps.

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug

```python
import lightning as L
import torch
import torch.nn.functional as F
from lightning.pytorch.demos import Transformer, WikiText2
from lightning.pytorch.callbacks import ModelCheckpoint, Callback
from torch.utils.data import DataLoader, random_split


class LanguageModel(L.LightningModule):
    def __init__(self, vocab_size):
        super().__init__()
        self.model = Transformer(vocab_size=vocab_size)
        self.model = torch.compile(self.model)

    def training_step(self, batch, batch_idx):
        input, target = batch
        output = self.model(input, target)
        loss = F.nll_loss(output, target.view(-1))
        self.log("train_loss", loss, prog_bar=True)
        return loss

    def validation_step(self, batch, batch_idx):
        input, target = batch
        output = self.model(input, target)
        loss = F.nll_loss(output, target.view(-1))
        self.log("val_loss", loss, prog_bar=True)
        return loss

    def test_step(self, batch, batch_idx):
        input, target = batch
        output = self.model(input, target)
        loss = F.nll_loss(output, target.view(-1))
        self.log("test_loss", loss, prog_bar=True)
        return loss

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.1)

    # def on_train_batch_end(self, outputs, batch, batch_idx):
    #     print(f"{self.trainer.global_step=}, {batch_idx=}")


class MyCallback(Callback):
    def on_train_batch_start(self, trainer, pl_module, batch, batch_idx):
        print(f"[TRAIN STEP START] {trainer.global_step=}, {batch_idx=}")

    def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx):
        print(f"[TRAIN STEP END] {trainer.global_step=}, {batch_idx=}")

    def on_validation_batch_start(self, trainer, pl_module, batch, batch_idx):
        print(f"[VAL STEP START] {trainer.global_step=}, {batch_idx=}")


class _ModelCheckpoint(ModelCheckpoint):
    """Modified version of ModelCheckpoint that saves the model after fit completes."""

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)


def main():
    L.seed_everything(42)

    # Data
    dataset = WikiText2()

    # Split data in to train, val, test
    n = len(dataset)
    train_dataset, val_dataset, test_dataset = random_split(
        dataset, [n - 200, 100, 100]
    )
    train_dataloader = DataLoader(train_dataset, batch_size=1, shuffle=False)
    val_dataloader = DataLoader(val_dataset, batch_size=1, shuffle=False)
    test_dataloader = DataLoader(test_dataset, batch_size=1, shuffle=False)
    print(f"{len(train_dataset)=}")

    # Model
    model = LanguageModel(vocab_size=dataset.vocab_size)

    # Callbacks
    checkpoint_callback = ModelCheckpoint(
        every_n_train_steps=5,
        save_top_k=-1,
        enable_version_counter=False,
        verbose=True,
    )
    my_callback = MyCallback()

    # Trainer
    trainer = L.Trainer(
        max_steps=10,
        val_check_interval=5,
        limit_val_batches=5,
        callbacks=[my_callback, checkpoint_callback],
        enable_progress_bar=False,
    )
    trainer.fit(model, train_dataloader, val_dataloader)
    # trainer.test(model, test_dataloader)

    # Resume training from checkpoint
    ckpt_path = "lightning_logs/version_0/checkpoints/epoch=0-step=5.ckpt"
    print(f"Resuming training from checkpoint {ckpt_path}")
    trainer.fit(
        model,
        train_dataloader,
        val_dataloader,
        ckpt_path=ckpt_path,
    )


if __name__ == "__main__":
    main()
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by DAKSH-coder26**
## What does this PR do?

This PR fixes incorrect synchronization of `batch_progress` when resuming training from a **mid-epoch checkpoint**, as described in issue **#19367**.

Before this fix, resuming from a checkpoint saved before completing an epoch caused:

- `batch_progress.total_ready` and `total_completed` to desynchronize from the restored `global_step`
- `current_ready` / `current_completed` to become inconsistent or invalid
- Validation, checkpoint intervals, or internal loop schedules to fire at the wrong steps
- Off-by-one errors in resumed training

### Summary of the Fix

The PR modifies `CheckpointConnector.resume_start()` to:

- Realign `batch_progress.total_ready` and `total_completed` to be at least the restored `global_step`
- Recompute `current_ready` and `current_completed` using  
  `global_step % limit_train_batches` when available
- Fall back safely when dataloader size or fields differ across Lightning versions
- Ensure behavior remains unchanged when no checkpoint is provided

A new test verifies correct behavior and prevents regressions.

### Tests Added

A new file:  
`tests/test_resume_batch_progress.py`

This test:

1. Trains to global_step = 5  
2. Saves a mid-epoch checkpoint  
3. Resumes from the checkpoint  
4. Ensures:
   - `batch_progress` exists in the expected form  
   - `total_completed` ≥ `global_step`  
   - `current_completed` ≤ `total_completed`  
   - All counters are non-negative and consistent  

The test **failed before the fix** and **passes after it**.

**Fixes #19367**

---

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this discussed/agreed via a GitHub issue? (Issue #19367)
- [x] Did you read the contributor guideline, Pull Request section?
- [x] Did you make sure your PR does only one thing?
- [x] No documentation changes were needed
- [x] Added the necessary tests
- [x] Verified new and existing tests pass locally
- [x] Listed breaking changes (none introduced)
- [x] Updated the CHANGELOG

</details>

---

## PR review

Anyone in the community is welcome to review the PR.

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review?
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Ensure the title and description clearly explain the PR
- [ ] Apply appropriate labels and milestones

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21411.org.readthedocs.build/en/21411/

### CODE REVIEW

**Comment by DAKSH-coder26** [comment]
Hi maintainers , this PR is ready for review from my side.

Some required checks (`schema`, `required-jobs`) are stuck in "Expected" due to the old Lit CI platform being shut down. 
Could you please trigger / approve the workflows or override the stuck checks?

Thanks!

File: src/lightning/pytorch/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -83,6 +83,40 @@ def resume_start(self, checkpoint_path: Optional[_PATH] = None, weights_only: Op
...
+
+                # Try to compute within-epoch counters from limit_train_batches when possible,
+                # otherwise fall back to safe defaults.
+                try:
+                    epoch_len = getattr(self.trainer, "limit_train_batches", None)
+                    if isinstance(epoch_len, int) and epoch_len > 0:
+                        epoch_size = int(epoch_len)
+                        bp.current_completed = gs % max(1, epoch_size)
+                        bp.current_ready = bp.current_completed
+                    else:
```
Comment (justusschock) [inline src/lightning/pytorch/trainer/connectors/checkpoint_connector.py]: do we need to set them to zero? aren't they zero by default?

File: src/lightning/pytorch/trainer/connectors/checkpoint_connector.py [hunk 2]
```diff
@@ -83,6 +83,40 @@ def resume_start(self, checkpoint_path: Optional[_PATH] = None, weights_only: Op
...
+                        bp.current_completed = 0
+                        bp.current_ready = 0
+                except Exception:
+                    try:
+                        bp.current_completed = 0
```
Comment (justusschock) [inline src/lightning/pytorch/trainer/connectors/checkpoint_connector.py]: why do we zero them here?

File: src/lightning/pytorch/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -83,6 +83,40 @@ def resume_start(self, checkpoint_path: Optional[_PATH] = None, weights_only: Op
             loaded_checkpoint = self.trainer.strategy.load_checkpoint(checkpoint_path, weights_only=weights_only)
         self._loaded_checkpoint = _pl_migrate_checkpoint(loaded_checkpoint, checkpoint_path)
 
+        try:
+            bp = getattr(self.trainer, "batch_progress", None)
+            if bp is not None:
+                gs = int(getattr(self.trainer, "global_step", 0))
+                # Align total counters defensively so they are at least the restored global_step
+                try:
+                    bp.total_ready = max(int(getattr(bp, "total_ready", 0)), gs)
```
Comment (justusschock) [inline src/lightning/pytorch/trainer/connectors/checkpoint_connector.py]: could you use more expressive variable names? like `batch_progress` instead of `bp`? makes it a bit easier to read :)

File: src/lightning/pytorch/trainer/connectors/checkpoint_connector.py (hunk 1 above)
Comment (DAKSH-coder26) [reply src/lightning/pytorch/trainer/connectors/checkpoint_connector.py]: Good catch — you’re right. The earlier version was redundantly zeroing these values
even though the defaults already handle this.

I’ve updated the logic to only set the counters when they can be derived from
`global_step` and `limit_train_batches`, and otherwise leave the defaults /
restored state untouched.

File: src/lightning/pytorch/trainer/connectors/checkpoint_connector.py (hunk 3 above)
Comment (DAKSH-coder26) [reply src/lightning/pytorch/trainer/connectors/checkpoint_connector.py]: Done 👍 Renamed `bp` to `batch_progress` and other locally introduced variables for clarity.

**Comment by DAKSH-coder26** [comment]
Hi! I’ve addressed the review comments and updated the PR accordingly. 
All conversations have been resolved. Whenever you have time, I’d appreciate a re-review. 
Thanks!
